### PR TITLE
DOC: Add examples for setbufsize & getbufsize

### PR DESCRIPTION
**Pull request by jnclt**
Add examples for getbufsize and setbufsize as requested by https://github.com/numpy/numpy/issues/22266.

Written as part of PyData 2022 sprint.

### CODE REVIEW

**Review by HaoZeke — CHANGES_REQUESTED** [review]
The results of these are dependent on the OS, and CI failures are indicative of this. `MacOS Python 3.8` [on Azure](https://dev.azure.com/numpy/numpy/_build/results?buildId=27038&view=logs&j=9d7fb089-bdae-57e1-a4ec-9025f62c646c&t=711c547c-351b-5827-f6c7-3aaedf13f9d4) reports:
```
numpy.core.setbufsize
---------------------

File "build/testenv/lib/python3.8/site-packages/numpy/__init__.py", line 199, in setbufsize
Failed example:
    np.getbufsize()
Expected:
    8192
Got:
    20000


File "build/testenv/lib/python3.8/site-packages/numpy/__init__.py", line 201, in setbufsize
Failed example:
    np.setbufsize(4096)
Expected:
    8192
Got:
    20000

numpy.core.getbufsize
---------------------

File "build/testenv/lib/python3.8/site-packages/numpy/__init__.py", line 232, in getbufsize
Failed example:
    np.getbufsize()
Expected:
    8192
Got:
    4096
```

**Comment by jnclt** [comment]
I didn't realize the examples are actually tested, thanks for pointing it out.
I don't see how to make them platform-independent, so I'd just close this PR.
Perhaps examples for these functions aren't that helpful after all?

**Comment by HaoZeke** [comment]
> I didn't realize the examples are actually tested, thanks for pointing it out. I don't see how to make them platform-independent, so I'd just close this PR. Perhaps examples for these functions aren't that helpful after all?

Rather than closing this perhaps just marking the outputs as "to be skipped", and mention the OS and indicative output in a comment?

```python
    refguide can be signalled to skip testing code by adding
    ``#doctest: +SKIP`` to the end of the line. If the output varies or is
    random, add ``# may vary`` or ``# random`` to the comment. for example

    >>> plt.plot(...)  # doctest: +SKIP
    >>> random.randint(0,10)
    5 # random
```

**Comment by mhvk** [comment]
I looked at this PR as I've often wondered whether the default buffer sizes are close to optimal for some of my more performance critical code, and was a little disappointed here to only see examples added of usage, which is really quite trivial (though good to mention it has to be a multiple of 16 bytes).

Maybe it is still useful, but that in that case probably best to set back the buffer size in the example -- as is, after the testing code passes through this example, the buffer size is now different for all further tests it does.

File: numpy/core/_ufunc_config.py [hunk 1]
```diff
@@ -182,13 +182,25 @@ def geterr():
...
+    setbufsize : int
+        Original size of ufunc buffer in bytes.
+
+    Examples
+    --------
+    >>> original_size = np.getbufsize()
+    >>> np.setbufsize(4096) == original_size
+    True
+    >>> np.getbufsize()
+    4096
```
Comment (HaoZeke) [inline numpy/core/_ufunc_config.py:203]:
```suggestion
    >>> original_size = np.getbufsize()
    >>> np.setbufsize(4096)
    >>> np.getbufsize()
    4096
    >>> np.setbufsize(original_size)
```

Comment (HaoZeke) [reply numpy/core/_ufunc_config.py:203]: The `==` test doesn't add anything and is potentially confusing. Also resets the size as @mhvk suggested.

File: numpy/core/_ufunc_config.py [hunk 2]
```diff
@@ -207,13 +219,17 @@ def setbufsize(size):
...
     Returns
     -------
     getbufsize : int
         Size of ufunc buffer in bytes.
-
+    
+    Examples
+    --------
+    >>> np.getbufsize()  # doctest: +SKIP
+    8192  # may vary
```
Comment (HaoZeke) [inline numpy/core/_ufunc_config.py:232]:
```suggestion
    8192  # varies by OS
```

**Review by HaoZeke — CHANGES_REQUESTED** [review]

File: numpy/core/_ufunc_config.py (hunk 2 above)
Comment (seberg) [reply numpy/core/_ufunc_config.py:232]: It doesn't vary by OS, the variations seem due to the test run itself, I think.  The typing tests are slow and don't run everwywhere, and set the bufsize to 4096 without setting it back.  The 20000 is/was maybe also the doc tests running twice or so?

I am not sure when to modify buffer sizes best.  There is an older issue I saw recently, that suggested we should just make them generally smaller, also to better use CPU caches (I guess times change)...

Comment (jnclt) [reply numpy/core/_ufunc_config.py:232]: I realize I don't know enough about the use cases of these functions to meaningfully improve the docs atm. If I manage to find an example that demonstrates the impact on performance, I'll open another PR. Feel free to close this one, unless you want to discuss this further.

**Comment by mattip** [comment]
i think this would be more convincing with guidance when to change the buffer size. Is there some explanation elsewhere as to why one would want to change it?

**Comment by seberg** [comment]
Its too bad, but I suppose we should close this eve though there are some nice starts smaller things in here.
Thanks for the attempt!  I suppose it wasn't really sprintable, because it needs at least some experimentation to add meaningful info (I made a note in the tracking issue).

If anyone wants to continue, please do not hesitate, or I am happy to reopen again.

(Which btw. could also lead to improving the buffer sizes ;)).

**Comment by InessaPawson** [comment]
@jnclt Thank you for giving it a go, Lefteris! Feel free to work on another [“sprintable” issue](https://github.com/numpy/numpy/issues?q=is%3Aopen+label%3Asprintable+sort%3Aupdated-desc) that peaks your interest.

**Pull request closed without merging**